### PR TITLE
mgr/cephadm: validating service_id for MDS

### PR DESCRIPTION
**Pull request by rkachach**
Fixes: https://tracker.ceph.com/issues/54184

Signed-off-by: Redouane Kachach <rkachach@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/orchestrator/module.py [hunk 1]
```diff
@@ -1074,6 +1074,9 @@ def _apply_mds(self,
             placement=PlacementSpec.from_string(placement),
             unmanaged=unmanaged,
             preview_only=dry_run)
+
+        spec.validate()  # force any validation exceptions to be caught correctly
```
Comment (mgfritch) [inline src/pybind/mgr/orchestrator/module.py:1078]: why is this needed?

File: src/python-common/ceph/deployment/service_spec.py [hunk 2]
```diff
@@ -680,7 +680,9 @@ def validate(self) -> None:
                         f'Service of type \'{self.service_type}\' should not contain a service id')
 
         if self.service_id:
-            if not re.match('^[a-zA-Z0-9_.-]+$', self.service_id):
+            if self.service_type == 'mds' and str(self.service_id)[0].isdigit():
+                raise SpecValidationError('MDS service id cannot start with a numeric digit')
+            elif not re.match('^[a-zA-Z0-9_.-]+$', self.service_id):
```
Comment (mgfritch) [inline src/python-common/ceph/deployment/service_spec.py:685]: should we consider having a well-defined `MDSSpec` class specifically for this new validation?

File: src/pybind/mgr/orchestrator/module.py (hunk 1 above)
Comment (rkachach) [reply src/pybind/mgr/orchestrator/module.py:1078]: The validate() method checks the spec object and validate its data to detect any error.

File: src/python-common/ceph/deployment/service_spec.py (hunk 2 above)
Comment (rkachach) [reply src/python-common/ceph/deployment/service_spec.py:685]: In my opinion current code is simple and easy to understand. In case we start adding more specific checks for special services then I'd also vote to go and create a dedicated sub-class for each service. I think we should also consider for the future introducing some kind of schema (jsonschema i.e) to validate this kind of data classes structure and fields.

File: src/pybind/mgr/orchestrator/module.py (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/orchestrator/module.py:1078]: it actually ends up reaching the validate anyway even without the validate call. I commented this line out and got

```
[ceph: root@vm-00 /]# ceph orch apply mds 5 
Error EINVAL: Traceback (most recent call last):
  File "/usr/share/ceph/mgr/mgr_module.py", line 1701, in _handle_command
    return self.handle_command(inbuf, cmd)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 171, in handle_command
    return dispatch[cmd['prefix']].call(self, cmd, inbuf)
  File "/usr/share/ceph/mgr/mgr_module.py", line 433, in call
    return self.func(mgr, **kwargs)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 107, in <lambda>
    wrapper_copy = lambda *l_args, **l_kwargs: wrapper(*l_args, **l_kwargs)  # noqa: E731
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 96, in wrapper
    return func(*args, **kwargs)
  File "/usr/share/ceph/mgr/orchestrator/module.py", line 1080, in _apply_mds
    return self._apply_misc([spec], dry_run, format, no_overwrite)
  File "/usr/share/ceph/mgr/orchestrator/module.py", line 1046, in _apply_misc
    raise_if_exception(completion)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 228, in raise_if_exception
    raise e
Exception: MDS service id cannot start with a numeric digit
```
which, in the cephadm mgr module actually went through

```
Traceback (most recent call last):
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 125, in wrapper
    return OrchResult(f(*args, **kwargs))
  File "/usr/share/ceph/mgr/cephadm/module.py", line 2504, in apply
    results.append(self._apply(spec))
  File "/usr/share/ceph/mgr/cephadm/module.py", line 2377, in _apply
    return self._apply_service_spec(cast(ServiceSpec, spec))
  File "/usr/share/ceph/mgr/cephadm/module.py", line 2482, in _apply_service_spec
    allow_colo=self.cephadm_services[spec.service_type].allow_colo(),
  File "/usr/share/ceph/mgr/cephadm/schedule.py", line 174, in validate
    self.spec.validate()
  File "/lib/python3.6/site-packages/ceph/deployment/service_spec.py", line 684, in validate
    raise SpecValidationError('MDS service id cannot start with a numeric digit')
ceph.deployment.hostspec.SpecValidationError: MDS service id cannot start with a numeric digit
```

so I think we don't actually need this call

File: src/python-common/ceph/deployment/service_spec.py (hunk 2 above)
Comment (adk3798) [reply src/python-common/ceph/deployment/service_spec.py]: there is actually precedent for putting fairly simple checks in validation functions for classes for service specs for specific services. E.g. the iscsi service spec has 

image

and alertmanager service spec has

image

so it's not out of the question to do something like that. While this is a fairly minor change, I think it's the only place in the generic service spec class that we're doing a special check for a certain service spec. For organizational reasons it could be worth creating an mds service spec class to be in line with how we've done it for other services. It would be super simple as well.

File: src/pybind/mgr/orchestrator/module.py (hunk 1 above)
Comment (rkachach) [reply src/pybind/mgr/orchestrator/module.py:1078]: we are already doing the same (validating the spec ASAP) in for the SNMPGatewaySpec:

```
        spec = SNMPGatewaySpec(
            snmp_version=snmp_version,
            port=port,
            credentials=credentials,
            snmp_destination=destination,
            engine_id=engine_id,
            auth_protocol=auth_protocol,
            privacy_protocol=privacy_protocol,
            placement=PlacementSpec.from_string(placement),
            unmanaged=unmanaged,
            preview_only=dry_run
        )

       spec.validate()  # force any validation exceptions to be caught correctly
```

In my opinion data (in general) have to be validated ASAP. In fact in the above trace we make several calls, in several classes to end up calling the validation (to detect that the spec is invalid). Something that we could have done earlier. In think both for code readability (and for analyzing errors in the future if any) it's better to do the validation ASAP so in this case I'd prefer to keep this explicit call to .validate() method.

File: src/python-common/ceph/deployment/service_spec.py (hunk 2 above)
Comment (rkachach) [reply src/python-common/ceph/deployment/service_spec.py]: I didn't notice this subclasses. Since we are already doing the same for other similar cases I agree to create a new subclass for MDS spec also. I'll change the code.

File: src/pybind/mgr/orchestrator/module.py (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/orchestrator/module.py:1078]: As discussed in standup, we're okay with having this check called directly, but maybe we should actually add this to other apply commands as well to have the functionality be uniform

File: src/pybind/mgr/orchestrator/module.py [hunk 3]
```diff
@@ -11,7 +11,7 @@
 from ceph.deployment.inventory import Device
 from ceph.deployment.drive_group import DriveGroupSpec, DeviceSelection, OSDMethod
 from ceph.deployment.service_spec import PlacementSpec, ServiceSpec, service_spec_allow_invalid_from_json, \
-    SNMPGatewaySpec
+    SNMPGatewaySpec, MDSSpec
```
Comment (adk3798) [inline src/pybind/mgr/orchestrator/module.py:14]: while I know SNMP is doing their import here, I think this should be taken from the ._interface file rather than the service_spec file directly to be in line with other specific service service spec imports. Honestly, I'd kind of like the SNMPGatewaySpec import to be from _interface as well.

Also, while we're talking about the _interface file, we should probably update the type signature of https://github.com/ceph/ceph/blob/master/src/pybind/mgr/orchestrator/_interface.py#L611 to take an mds spec the same way rgw, ingress etc. below it take their specific spec.

File: src/pybind/mgr/orchestrator/module.py (hunk 1 above)
Comment (rkachach) [reply src/pybind/mgr/orchestrator/module.py:1077]: I added validate() checks for the other services

File: src/pybind/mgr/orchestrator/module.py (hunk 3 above)
Comment (rkachach) [reply src/pybind/mgr/orchestrator/module.py]: Code has been updated accordingly. New code import SNMPGatewaySpec and MDSSpec from the interface.

File: src/pybind/mgr/orchestrator/module.py [hunk 4]
```diff
@@ -1115,6 +1114,8 @@ def _apply_rgw(self,
             preview_only=dry_run
         )
 
+        spec.validate()  # force any validation exceptions to be caught correctly
```
Comment (phlogistonjohn) [inline src/pybind/mgr/orchestrator/module.py:1117]: This is minor but do you mean "early" or "immediately" rather than "correctly" here?

Comment (rkachach) [reply src/pybind/mgr/orchestrator/module.py:1117]: in this case is immediatly since the exception will be raised right away

File: src/pybind/mgr/orchestrator/module.py (hunk 1 above)
Comment (mgfritch) [reply src/pybind/mgr/orchestrator/module.py:1077]: see also: https://github.com/ceph/ceph/pull/44995

File: src/python-common/ceph/deployment/service_spec.py [hunk 5]
```diff
@@ -1299,3 +1300,27 @@ def validate(self) -> None:
...
 yaml.add_representer(SNMPGatewaySpec, ServiceSpec.yaml_representer)
+
+
+class MDSSpec(ServiceSpec):
+    def __init__(self,
+                 service_type: str = 'mds',
+                 service_id: Optional[str] = None,
+                 placement: Optional[PlacementSpec] = None,
+                 unmanaged: bool = False,
+                 preview_only: bool = False,
```
Comment (adk3798) [inline src/python-common/ceph/deployment/service_spec.py:1311]: @mgfritch do you think there are any other parameters we could maybe need for mds or is this enough

**Review by adk3798 — APPROVED** [review]

**Comment by rkachach** [comment]
jenkins retest this please

File: src/pybind/mgr/orchestrator/_interface.py [hunk 6]
```diff
@@ -608,7 +608,7 @@ def apply_mgr(self, spec: ServiceSpec) -> OrchResult[str]:
         """Update mgr cluster"""
         raise NotImplementedError()
 
-    def apply_mds(self, spec: ServiceSpec) -> OrchResult[str]:
+    def apply_mds(self, spec: MDSSpec) -> OrchResult[str]:
```
Comment (adk3798) [inline src/pybind/mgr/orchestrator/_interface.py:611]:
```
mds_autoscaler/__init__.py:6: note: In module imported here:
mds_autoscaler/module.py: note: In member "verify_and_manage_mds_instance" of class "MDSAutoscaler":
mds_autoscaler/module.py:83: error: Argument 1 to "apply_mds" of "Orchestrator" has incompatible type "ServiceSpec"; expected "MDSSpec"
Found 1 error in 1 file (checked 31 source files)
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/.tox/mypy/bin/mypy --config-file=../../mypy.ini -m alerts -m balancer -
```
we need a small type change here to make mypy happy

**Review by adk3798 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/orchestrator/_interface.py (hunk 6 above)
Comment (adk3798) [reply src/pybind/mgr/orchestrator/_interface.py:611]: or well, I guess I should say we need a small change in that autoscaler module, not here

Comment (rkachach) [reply src/pybind/mgr/orchestrator/_interface.py:611]: @adk3798 I fixed it in the mds_autoscaler module

**Comment by adk3798** [comment]
http://pulpito.ceph.com/adking-2022-02-17_13:37:32-orch:cephadm-wip-adk-testing-2022-02-17-0053-distro-basic-smithi/

Failures tracked by https://tracker.ceph.com/issues/54304

File: src/pybind/mgr/mds_autoscaler/module.py [hunk 7]
```diff
@@ -32,12 +32,12 @@ def get_service(self, fs_name: str) -> Optional[orchestrator.ServiceDescription]
             return completion.result[0]
         return None
 
-    def update_daemon_count(self, spec: ServiceSpec, fs_name: str, abscount: int) -> ServiceSpec:
+    def update_daemon_count(self, spec: MDSSpec, fs_name: str, abscount: int) -> MDSSpec:
```
Comment (adk3798) [inline src/pybind/mgr/mds_autoscaler/module.py:35]: Still getting mypy failures causing make check to fail. I think we need to cast from service spec to mds spec here https://github.com/ceph/ceph/blob/master/src/pybind/mgr/mds_autoscaler/module.py#L82

**Review by adk3798 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/mds_autoscaler/module.py (hunk 7 above)
Comment (rkachach) [reply src/pybind/mgr/mds_autoscaler/module.py]: I fixed the signature of update_daemon_count so now it uses ServiceSpec instead of MDSSpec:

`def update_daemon_count(self, spec: ServiceSpec, fs_name: str, abscount: int) -> MDSSpec:`

this fixes the mypy error

**Review by adk3798 — APPROVED** [review]
